Extra notes on parallelization efficiencies (#1046)
blue-yonder/tsfresh: docs/text/tsfresh_on_a_cluster.rst

To construct your custom Distributor, you need to define an object that inherits from the abstract base class
:class:`tsfresh.utilities.distribution.DistributorBaseClass`.
The :mod:`tsfresh.utilities.distribution` module contains more information about what you need to implement.

Notes for efficient parallelization
'''''''''''''''''''''''''''''''''''

By default tsfresh uses parallelization to distribute the single threaded python code out to make use of the multiple threads/cores available on the host machine.

> However, this can create an issue known as over-provisioning, due to the fact that many of the underlying python libraries this project is built upon drop back into C code implementations for their low-level processing, and `also` try and spread their workload between as many threads/cores available.

Reviewer: ```suggestion
However, this can create an issue known as over-provisioning. Many of the underlying python libraries (e.g. numpy) used in the feature calculators have C code implementations for their low-level processing. Those `also` try to spread their workload between as many cores available - which is in conflict with the parallelization done by tsfresh.
```

Over-provisioning is inefficient because of the overheads of repeated context switching. 

> It is far better to give control over this load distribution to the larger tasks, hence using parallelization in tsfresh, but to make best use of this and stop the over-provisioning of cpu cores it is required to prevent the low-level libraries attempting to distribute their workload over multiple threads/cores.

Reviewer: I think you can drop this paragraph: you described very nicely before that over provisioning is not what we want!

> This can be achieved by setting some environment variables early on in the kernel before these underlying python modules get loaded in.

Reviewer: ```suggestion
This issue can be solved by constraining the C libraries to single threads, using the following environment variables:
```

> The environment variables in question are, `OMP_NUM_THREADS`, `MKL_NUM_THREADS` and `OPENBLAS_NUM_THREADS`. All of these should be set to `1`. For example, if using a Jupyterlab environment, having the first element of the notebook as the following will achieve this:

Reviewer: I think you can remove this paragraph as you use them in the code example

.. code:: python

    import os

>     os.environ['OMP_NUM_THREADS']="1"
>     os.environ['MKL_NUM_THREADS']="1"
>     os.environ['OPENBLAS_NUM_THREADS']="1"

Reviewer: ```suggestion
    os.environ['OMP_NUM_THREADS'] = "1"
    os.environ['MKL_NUM_THREADS'] = "1"
    os.environ['OPENBLAS_NUM_THREADS'] = "1"
```

> To run the notebook successfully it is then required to use the 'Restart the kernel' option.

Reviewer: ```suggestion
Put these lines at the beginning of your notebook/python script - before you call any tsfresh code.
```
Is it really required to restart the notebook again - even if those are the first lines? I guess not?

> The more cores your host computer has, the more improvement in processing speed will be gained by implementing these environment changes. Speed increases of between 6x and 26x have been observed depending on the class of the host machine.

Reviewer: ```suggestion
The more cores your host computer has, the more improvement in processing speed will be gained by implementing these environment changes. Speed increases of between 6x and 26x have been observed depending on the type of the host machine.
```
